crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+    }
+}
+
+async fn resolve_transaction(
+    State(state): State<RestService>,
+    accept: AcceptFormat,
+    Json(unresolved_transaction): Json<UnresolvedTransaction>,
+) -> Result<ResponseContent<ResolveTransactionResponse>> {
+    let executor = state
+        .executor
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("No Transaction Executor"))?;
+    let (reference_gas_price, protocol_config) = {
+        let system_state = state.reader.get_system_state_summary()?;
+
+        let current_protocol_version = state.reader.get_system_state_summary()?.protocol_version;
+
+        let protocol_config = ProtocolConfig::get_for_version_if_supported(
+            current_protocol_version.into(),
+            state.reader.inner().get_chain_identifier()?.chain(),
+        )
+        .ok_or_else(|| {
+            RestError::new(
+                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                "unable to get current protocol config",
+            )
+        })?;
+
+        (system_state.reference_gas_price, protocol_config)
+    };
+    let called_packages =
+        called_packages(&state.reader, &protocol_config, &unresolved_transaction)?;
+    let user_provided_budget = unresolved_transaction
+        .gas_payment
+        .as_ref()
+        .and_then(|payment| payment.budget);
+    let mut resolved_transaction = resolve_object_references(
+        &state.reader,
+        &called_packages,
+        reference_gas_price,
+        protocol_config.max_tx_gas(),
+        unresolved_transaction,
+    )?;
+
+    let simulation_result = executor
+        .simulate_transaction(resolved_transaction.clone())
+        .map_err(anyhow::Error::from)?;
+
+    let budget = if let Some(user_provided_budget) = user_provided_budget {
+        user_provided_budget
+    } else {
+        let estimate = estimate_gas_budget_from_gas_cost(
+            simulation_result.effects.gas_cost_summary(),
+            reference_gas_price,
+        );
+        resolved_transaction.gas_data_mut().budget = estimate;
+        estimate
+    };
+
+    // If we ended up using a mock gas coin we need to do gas selection now
+    if simulation_result.mock_gas_id.is_some() {
+        let input_objects = resolved_transaction
+            .input_objects()
+            .map_err(anyhow::Error::from)?
+            .iter()
+            .flat_map(|obj| match obj {
+                sui_types::transaction::InputObjectKind::ImmOrOwnedMoveObject((id, _, _)) => {
+                    Some(*id)
+                }
+                _ => None,
+            })
+            .collect_vec();
+        let gas_coins = select_gas(
+            &state.reader,
+            resolved_transaction.gas_data().owner,
+            budget,
+            protocol_config.max_gas_payment_objects(),
+            &input_objects,
+        )?;
+        resolved_transaction.gas_data_mut().payment = gas_coins;
+    }
+
+    // Do one last dry-run
```

Comment:
Can we make this last dry-run optional as well? In GraphQL we would only ask for it, if the user asked for simulation results.